### ISSUE TITLE
Build issue on certain MSVC in Cantera.

### ISSUE THREAD

**Issue by 1uc** [BlueBrain/HighFive#724]
The CI of https://github.com/Cantera uncovered a build issue for certain versions of MSVC. Some of the discussion started in #713. 

The link to the failing CI is:
https://github.com/Cantera/cantera/actions/runs/4588940048/jobs/8103492434

It reports a
```
c:\miniconda3\envs\test\library\include\highfive\bits/H5Path_traits_misc.hpp(31): error C2248: 'HighFive::Object::Object': cannot access protected member declared in class 'HighFive::Object'
...
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.16.27023\include\filesystem(2392): note: see reference to class template instantiation 'std::chrono::time_point<std::filesystem::_File_time_clock,std::filesystem::_File_time_clock::duration>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.16.27023\include\type_traits(616): note: see reference to class template instantiation 'std::basic_string_view<wchar_t,std::char_traits<wchar_t>>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.16.27023\include\xstring(2124): note: see reference to class template instantiation 'std::is_convertible<const _StringViewIsh &,std::basic_string_view<wchar_t,std::char_traits<wchar_t>>>' being compiled
```
Which smell like SFINAE to check if something HighFive related can be converted to a string. While doing so it fails because the ctor of `HighFive::Object` is protected. To be investigated further.

**Comment by ischoegl** [comment]
Thanks, @1uc ... trying to debug a couple of other quirks, I updated/fixed our build scripts (unfortunately we're using a different build system - `SCons` rather than `cmake`) and I now can follow installation details a little better. The offending issue is now bypassed, as our installation now falls back to an earlier git submodule 2.6.2. 

The latest runner we're using does have a system installation of 2.7.0 from conda, which produces the same output as above for MSVC 14.1:
```
c:\miniconda3\envs\test\library\include\highfive\bits/H5Path_traits_misc.hpp(31): error C2248: 'HighFive::Object::Object': cannot access protected member declared in class 'HighFive::Object'
c:\miniconda3\envs\test\library\include\highfive\bits/H5Object_misc.hpp(27): note: see declaration of 'HighFive::Object::Object'
c:\miniconda3\envs\test\library\include\highfive\H5Object.hpp(53): note: see declaration of 'HighFive::Object'
c:\miniconda3\envs\test\library\include\highfive\bits/H5Path_traits_misc.hpp(19): note: while compiling class template member function 'HighFive::PathTraits<HighFive::Attribute>::PathTraits(void)'
c:\miniconda3\envs\test\library\include\highfive\bits\../H5Attribute.hpp(124): note: see reference to function template instantiation 'HighFive::PathTraits<HighFive::Attribute>::PathTraits(void)' being compiled
c:\miniconda3\envs\test\library\include\highfive\bits\../H5Attribute.hpp(45): note: see reference to class template instantiation 'HighFive::PathTraits<HighFive::Attribute>' being compiled 
```

**Comment by 1uc** [comment]
I have a suspicion why this is. We (in order to appease a compiler) removed some friend declarations in this commit:
https://github.com/BlueBrain/HighFive/commit/20be06fe1882cdd2423bc72cf7adbec32c529c24

Essentially, we're avoiding a few of these
```
class Foo {
    // ...
    template <typename Derivate>
    friend class PathTraits;
};
```

We should try to put them back conditionally for MSVC only.

**Comment by ischoegl** [comment]
Newer MSVC versions (2019 and later, i.e. 14.2 and 14.3) work as far as I can tell. MSVC toolset 14.1 is the one with issues on our end, but I'm not able to test older versions.

**Comment by 1uc** [comment]
@ischoegl It seems to be this compiler bug:
https://developercommunity.visualstudio.com/t/MSVC-compiler-improperly-implements-N489/1516410

If we look at `HighFive::File` we find:
```
class File: public Object, public NodeTraits<File>, public AnnotateTraits<File> {
  // ... 

  protected:
    File() = default;

  private:
    using Object::Object;
};
```
The error doesn't manifest with `--std=c++14` but only with `--std=c++17` (just like in the description of the bug). We're using VS 2022 17.0; the bug report doesn't specify the compiler version.

**Comment by ischoegl** [comment]
Confirming that we’re using the `c++17` flag also; I’m happy that you can reproduce outside of our CI. Based on this being a compiler bug that’s unlikely to be fixed (under investigation since ‘21) putting in a dedicated flag is a tough call. Since HighFive is an optional dependency for us at the moment (we just switched over from `h5py` and only use it in the development version), it’s easy to address on our end.

**Comment by 1uc** [comment]
I think you beat me to it, but yes, we can make the issue go away:
https://github.com/BlueBrain/HighFive/actions/runs/4611730629/jobs/8151778661?pr=728

I'll discuss with the other developers to see if we want to keep the workaround.

**Comment by ischoegl** [comment]
:+1: ... fwiw, I *guess* it's a function on how many people will try to use MSVC toolset 14.1 together with the `c++17` option. We use it as it is presumably the oldest version that should fully support the standard (I believe after revision 14.14). 

As an aside, it would be interesting to see whether a GH `runs-on: windows-2019` Action will produce the same behavior, apart from the open question whether the next VS will fix it. Looking over #727  / #728, restricting the workaround to the buggy MSVC toolset and not exposing `HIGHFIVE_HAS_FRIENDS` as a cmake flag may be an option. Regardless of the outcome, I'm truly impressed by the project - while I spent some time debugging, your responsiveness and support are truly appreciated!

**Comment by 1uc** [comment]
We've merged #728 into `master`. I tend to regret hard-coding anything build-system related; and since nobody is expected to set the variable, it acts pure as a safety valve that allows users to pick. Hopefully, it wont be used.

The name of the macro changed to `HIGHFIVE_HAS_FRIEND_DECLARATIONS` (or similar).

**Comment by 1uc** [comment]
@ischoegl Version 2.7.1 has been created. I'll close the issue since, most likely, it's been fixed. Please feel free to re-open or open a different issue as needed.

**Comment by ischoegl** [comment]
Thanks! While I'm relatively certain that all issues are resolved, I'll have to wait for the 2.7.1 conda packages to have final confirmation. Again, the HighFive team's efforts are truly appreciated!
